aclk/aclk_lws_wss_client.c

```diff
@@ -0,0 +1,315 @@
+	struct lws_wss_packet_buffer *data;
+	int retval = 0;
+
+	if( !inst ) {
+		error("Callback received without any aclk_lws_wss_engine_instance!");
+		return -1;
+	}
+
+	if( inst->upstream_reconnect_request ) {
+		error("Closing lws connectino due to libmosquitto error.");
+		char *upstream_connection_error = "MQTT protocol error. Closing underlying wss connection.";
+		lws_close_reason(wsi, LWS_CLOSE_STATUS_PROTOCOL_ERR, upstream_connection_error, strlen(upstream_connection_error));
+		retval = -1;
+		inst->upstream_reconnect_request = 0;
+	}
+
+	switch (reason) {
+	case LWS_CALLBACK_CLIENT_WRITEABLE:
+		aclk_lws_mutex_lock(&inst->write_buf_mutex);
+		data = lws_wss_packet_buffer_pop(&inst->write_buffer_head);
+		if(likely(data)) {
+			lws_write(wsi, data->data + LWS_PRE, data->data_size, LWS_WRITE_BINARY);
+			lws_wss_packet_buffer_free(data);
+			if(inst->write_buffer_head)
+				lws_callback_on_writable(inst->lws_wsi);
+		}
+		aclk_lws_mutex_unlock(&inst->write_buf_mutex);
+		break;
+	case LWS_CALLBACK_CLIENT_RECEIVE:
+		aclk_lws_mutex_lock(&inst->read_buf_mutex);
+		if(!received_data_to_ringbuff(inst->read_ringbuffer, in, len))
+			retval = 1;
+		aclk_lws_mutex_unlock(&inst->read_buf_mutex);
+
+		if(likely(inst->callbacks.data_rcvd_callback))
+			// to future myself -> do not call this while read lock is active as it will eventually
+			// want to acquire same lock later in aclk_lws_wss_client_read() function
+			inst->callbacks.data_rcvd_callback();
+		else
+			inst->data_to_read = 1; //to inform logic above there is reason to call mosquitto_loop_read
+		break;
+	case LWS_CALLBACK_PROTOCOL_INIT:
+		//initial connection here
+		//later we will reconnect with delay od ACLK_LWS_WSS_RECONNECT_TIMEOUT
+		//in case this connection fails or drops
+		_aclk_wss_connect(inst);
+		break;
+	case LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED:
+		//TODO if already active make some error noise
+		//currently we expect only one connection per netdata
+		inst->lws_wsi = wsi;
+		break;
+#ifdef AUTO_RECONNECT_ON_LWS_LAYER
+	case LWS_CALLBACK_USER:
+		inst->reconnect_timeout_running = 0;
+		_aclk_wss_connect(inst);
+		break;
+#endif
+	case LWS_CALLBACK_CLIENT_CLOSED:
+	case LWS_CALLBACK_WS_PEER_INITIATED_CLOSE:
+	case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
+#ifdef AUTO_RECONNECT_ON_LWS_LAYER
+		if(!inst->reconnect_timeout_running) {
+			lws_timed_callback_vh_protocol(lws_get_vhost(wsi),
+						   lws_get_protocol(wsi),
+					       LWS_CALLBACK_USER, ACLK_LWS_WSS_RECONNECT_TIMEOUT);
+			inst->reconnect_timeout_running = 1;
+		}
+#endif
+		//no break here on purpose we want to continue with LWS_CALLBACK_WSI_DESTROY
+	case LWS_CALLBACK_WSI_DESTROY:
+		aclk_lws_wss_clear_io_buffers(inst);
+		inst->lws_wsi = NULL;
+		inst->websocket_connection_up = 0;
+		break;
+	case LWS_CALLBACK_CLIENT_ESTABLISHED:
+		inst->websocket_connection_up = 1;
+		if(inst->callbacks.connection_established_callback)
+			inst->callbacks.connection_established_callback();
+		break;
```

Comment:
Should we not add a `default` to avoid compilation warnings?

Reply:
I already saw compilers doing this in the past.

Reply:
will add default to switch